File: tests/test_transform_effects.cpp
#include "utils.h"

#include <opentimelineio/clip.h>
#include <opentimelineio/deserialization.h>
#include <opentimelineio/externalReference.h>
#include <opentimelineio/transformEffects.h>

namespace otime = opentime::OPENTIME_VERSION;
namespace otio  = opentimelineio::OPENTIMELINEIO_VERSION;

int
main(int argc, char** argv)
{
    Tests tests;
    tests.add_test("test_video_transform_read", [] {
        using namespace otio;

        otio::ErrorStatus              status;
        SerializableObject::Retainer<> so =
            SerializableObject::from_json_string(
                R"(
            {
                "OTIO_SCHEMA": "Clip.1",
                "media_reference": {
                    "OTIO_SCHEMA": "ExternalReference.1",
                    "target_url": "unit_test_url",
                    "available_range": {
                        "OTIO_SCHEMA": "TimeRange.1",
                        "duration": {
                            "OTIO_SCHEMA": "RationalTime.1",
                            "rate": 24,
                            "value": 8
                        },
                        "start_time": {
                            "OTIO_SCHEMA": "RationalTime.1",
                            "rate": 24,
                            "value": 10
                        }
                    }
                },
                "effects": [
                    {
                        "OTIO_SCHEMA": "VideoScale.1",
                        "name": "scale",
                        "width": "1/2",
                        "height": "1/2",
                        "effect_name": "VideoScale",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoPosition.1",
                        "name": "position",
                        "x": "-1/2",
                        "y": "-1/2",
                        "effect_name": "VideoPosition",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoRotate.1",
                        "name": "rotate",
                        "rotation": "90/360",
                        "effect_name": "VideoRotate",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoCrop.1",
                        "name": "crop",
                        "left": "-1/4",
                        "right": "1/4",
                        "top": "-1/4",
                        "bottom": "1/4",
                        "effect_name": "VideoCrop",
                        "enabled": true
                    }
                ]
            })",
                &status);

        assertFalse(is_error(status));

        const Clip* clip = dynamic_cast<const Clip*>(so.value);
        assertNotNull(clip);

        auto effects = clip->effects();
        assertEqual(effects.size(), 4);

        auto video_scale = dynamic_cast<const VideoScale*>(effects[0].value);
        assertNotNull(video_scale);
        assertEqual(video_scale->width(), Rational(1, 2));
        assertEqual(video_scale->height(), Rational(1, 2));

        auto video_position = dynamic_cast<const VideoPosition*>(effects[1].value);
        assertNotNull(video_position);
        assertEqual(video_position->x(), Rational(-1, 2));
        assertEqual(video_position->y(), Rational(-1, 2));

        auto video_rotate = dynamic_cast<const VideoRotate*>(effects[2].value);
        assertNotNull(video_rotate);
        assertEqual(video_rotate->rotation(), Rational(90, 360));

        auto video_crop = dynamic_cast<const VideoCrop*>(effects[3].value);
        assertNotNull(video_crop);
        assertEqual(video_crop->left(), Rational(-1, 4));
        assertEqual(video_crop->right(), Rational(1, 4));
        assertEqual(video_crop->top(), Rational(-1, 4));
        assertEqual(video_crop->bottom(), Rational(1, 4));
    });

    tests.add_test("test_video_transform_write", [] {
        using namespace otio;

        SerializableObject::Retainer<otio::Clip> clip(new otio::Clip(
            "unit_clip",
            new otio::ExternalReference("unit_test_url"),
            std::nullopt,
            otio::AnyDictionary(),
            { new otio::VideoScale("scale", otime::Rational(1, 2), otime::Rational(1, 2)),
              new otio::VideoPosition("position", otime::Rational(-1, 2), otime::Rational(-1, 2)),
              new otio::VideoRotate("rotate", otime::Rational(90, 360)),
              new otio::VideoCrop("crop", otime::Rational(-1, 4), otime::Rational(1, 4), otime::Rational(-1, 4), otime::Rational(1, 4)) }));

        auto json = clip.value->to_json_string();

        std::string expected_json = R"({
    "OTIO_SCHEMA": "Clip.2",
    "metadata": {},
    "name": "unit_clip",
    "source_range": null,
    "effects": [
        {
            "OTIO_SCHEMA": "VideoScale.1",
            "metadata": {},
            "name": "scale",
            "effect_name": "VideoScale",
            "enabled": true,
            "width": "1/2",
            "height": "1/2"
        },
        {
            "OTIO_SCHEMA": "VideoPosition.1",
            "metadata": {},
            "name": "position",
            "effect_name": "VideoPosition",
            "enabled": true,
            "x": "-1/2",
            "y": "-1/2"
        },
        {
            "OTIO_SCHEMA": "VideoRotate.1",
            "metadata": {},
            "name": "rotate",
            "effect_name": "VideoRotate",
            "enabled": true,
            "rotation": "1/4"
        },
        {
            "OTIO_SCHEMA": "VideoCrop.1",
            "metadata": {},
            "name": "crop",
            "effect_name": "VideoCrop",
            "enabled": true,
            "left": "-1/4",
            "right": "1/4",
            "top": "-1/4",
            "bottom": "1/4"
        }
    ],
    "markers": [],
    "enabled": true,
    "media_references": {
        "DEFAULT_MEDIA": {
            "OTIO_SCHEMA": "ExternalReference.1",
            "metadata": {},
            "name": "",
            "available_range": null,
            "available_image_bounds": null,
            "target_url": "unit_test_url"
        }
    },
    "active_media_reference_key": "DEFAULT_MEDIA"
})";

        assertEqual(json, expected_json);

    });

    tests.run(argc, argv);
    return 0;
}

File: tests/test_transform_effects.py
"""Transform effects class test harness."""

import unittest
from fractions import Fraction

import opentimelineio as otio
import opentimelineio.test_utils as otio_test_utils

class VideoScaleTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        scale = otio.schema.VideoScale(
            name="ScaleIt",
            width=Fraction(1, 2),
            height=Fraction(2, 1),
            metadata={
                "foo": "bar"
            }
        )
        self.assertEqual(scale.width, Fraction(1, 2))
        self.assertEqual(scale.height, Fraction(2, 1))
        self.assertEqual(scale.name, "ScaleIt")
        self.assertEqual(scale.metadata, {"foo": "bar"})

    def test_eq(self):
        scale1 = otio.schema.VideoScale(
            name="ScaleIt",
            width=Fraction(1, 2),
            height=Fraction(2, 1),
            metadata={
                "foo": "bar"
            }
        )
        scale2 = otio.schema.VideoScale(
            name="ScaleIt",
            width=Fraction(1, 2),
            height=Fraction(2, 1),
            metadata={
                "foo": "bar"
            }
        )
        self.assertIsOTIOEquivalentTo(scale1, scale2)

    def test_serialize(self):
        scale = otio.schema.VideoScale(
            name="ScaleIt",
            width=Fraction(1, 2),
            height=Fraction(2, 1),
            metadata={
                "foo": "bar"
            }
        )
        encoded = otio.adapters.otio_json.write_to_string(scale)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(scale, decoded)

    def test_setters(self):
        scale = otio.schema.VideoScale(
            name="ScaleIt",
            width=Fraction(1, 2),
            height=Fraction(2, 1),
            metadata={
                "foo": "bar"
            }
        )
        self.assertEqual(scale.width, Fraction(1, 2))
        scale.width = 0.25
        self.assertEqual(scale.width, Fraction(1, 4))

class VideoCropTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_cons(self):
        crop = otio.schema.VideoCrop(
            name="CropIt",
            left=Fraction(1, 10),
            right=Fraction(2, 10),
            top=Fraction(3, 10),
            bottom=Fraction(4, 10),
            metadata={
                "baz": "qux"
            }
        )
        self.assertEqual(crop.left, Fraction(1, 10))
        self.assertEqual(crop.right, Fraction(2, 10))
        self.assertEqual(crop.top, Fraction(3, 10))
        self.assertEqual(crop.bottom, Fraction(4, 10))
        self.assertEqual(crop.name, "CropIt")
        self.assertEqual(crop.metadata, {"baz": "qux"})

    def test_eq(self):
        crop1 = otio.schema.VideoCrop(
            name="CropIt",
            left=Fraction(1, 10),
            right=Fraction(2, 10),
            top=Fraction(3, 10),
            bottom=Fraction(4, 10),
            metadata={
                "baz": "qux"
            }
        )
        crop2 = otio.schema.VideoCrop(
            name="CropIt",
            left=Fraction(1, 10),
            right=Fraction(2, 10),
            top=Fraction(3, 10),
            bottom=Fraction(4, 10),
            metadata={
                "baz": "qux"
            }
        )
        self.assertIsOTIOEquivalentTo(crop1, crop2)

    def test_serialize(self):
        crop = otio.schema.VideoCrop(
            name="CropIt",
            left=Fraction(1, 10),
            right=Fraction(2, 10),
            top=Fraction(3, 10),
            bottom=Fraction(4, 10),
            metadata={
                "baz": "qux"
            }
        )
        encoded = otio.adapters.otio_json.write_to_string(crop)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(crop, decoded)

    def test_setters(self):
        crop = otio.schema.VideoCrop(
            name="CropIt",
            left=Fraction(1, 10),
            right=Fraction(2, 10),
            top=Fraction(3, 10),
            bottom=Fraction(4, 10),
            metadata={
                "baz": "qux"
            }
        )
        self.assertEqual(crop.left, Fraction(1, 10))
        crop.left = 0.2
        self.assertEqual(crop.left, Fraction(1, 5))
        crop.right = 0.3
        self.assertEqual(crop.right, Fraction(3, 10))
        crop.top = 0.4
        self.assertEqual(crop.top, Fraction(2, 5))
        crop.bottom = 0.5
        self.assertEqual(crop.bottom, Fraction(1, 2))

class VideoPositionTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        position = otio.schema.VideoPosition(
            name="PositionIt",
            x=Fraction(3, 4),
            y=Fraction(5, 6),
            metadata={
                "alpha": "beta"
            }
        )
        self.assertEqual(position.x, Fraction(3, 4))
        self.assertEqual(position.y, Fraction(5, 6))
        self.assertEqual(position.name, "PositionIt")
        self.assertEqual(position.metadata, {"alpha": "beta"})

    def test_eq(self):
        pos1 = otio.schema.VideoPosition(
            name="PositionIt",
            x=Fraction(3, 4),
            y=Fraction(5, 6),
            metadata={
                "alpha": "beta"
            }
        )
        pos2 = otio.schema.VideoPosition(
            name="PositionIt",
            x=Fraction(3, 4),
            y=Fraction(5, 6),
            metadata={
                "alpha": "beta"
            }
        )
        self.assertIsOTIOEquivalentTo(pos1, pos2)

    def test_serialize(self):
        position = otio.schema.VideoPosition(
            name="PositionIt",
            x=Fraction(3, 4),
            y=Fraction(5, 6),
            metadata={
                "alpha": "beta"
            }
        )
        encoded = otio.adapters.otio_json.write_to_string(position)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(position, decoded)

    def test_setters(self):
        position = otio.schema.VideoPosition(
            name="PositionIt",
            x=Fraction(3, 4),
            y=Fraction(5, 6),
            metadata={
                "alpha": "beta"
            }
        )
        self.assertEqual(position.x, Fraction(3, 4))
        position.x = 0.25
        self.assertEqual(position.x, Fraction(1, 4))
        position.y = 0.5
        self.assertEqual(position.y, Fraction(1, 2))

class VideoRotateTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        rotate = otio.schema.VideoRotate(
            name="RotateIt",
            rotation=Fraction(90, 360),
            metadata={
                "rot": "val"
            }
        )
        self.assertEqual(rotate.rotation, Fraction(90, 360))
        self.assertEqual(rotate.name, "RotateIt")
        self.assertEqual(rotate.metadata, {"rot": "val"})

    def test_eq(self):
        rot1 = otio.schema.VideoRotate(
            name="RotateIt",
            rotation=Fraction(90, 360),
            metadata={
                "rot": "val"
            }
        )
        rot2 = otio.schema.VideoRotate(
            name="RotateIt",
            rotation=Fraction(90, 360),
            metadata={
                "rot": "val"
            }
        )
        self.assertIsOTIOEquivalentTo(rot1, rot2)

    def test_serialize(self):
        rotate = otio.schema.VideoRotate(
            name="RotateIt",
            rotation=Fraction(90, 360),
            metadata={
                "rot": "val"
            }
        )
        encoded = otio.adapters.otio_json.write_to_string(rotate)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(rotate, decoded)

    def test_setters(self):
        rotate = otio.schema.VideoRotate(
            name="RotateIt",
            rotation=Fraction(90, 360),
            metadata={
                "rot": "val"
            }
        )
        self.assertEqual(rotate.rotation, Fraction(90, 360))
        rotate.rotation = Fraction(45, 360)
        self.assertEqual(rotate.rotation, Fraction(45, 360))
        rotate.rotation = Fraction(22, 360)
        self.assertEqual(rotate.rotation, Fraction(22, 360))
